File: sample/binary-search/lis.go
// Longest Increasing Subsequence O(NlogN)
package main

import "fmt"

func lis(a []int) int {
	dp := []int{}
	for i := 0; i < len(a); i++ {
		if len(dp) == 0 || dp[len(dp)-1] < a[i] {
			dp = append(dp, a[i])
		} else {
			lo, rt, hi := 0, 0, len(dp)-1
			for lo <= hi {
				mi := lo + (hi-lo)/2
				if a[i] <= dp[mi] {
					hi = mi - 1
					rt = mi
				} else {
					lo = mi + 1
				}
			}
			dp[rt] = a[i]
		}
	}
	return len(dp)
}

func main() {
	fmt.Println(lis([]int{3, 2, 5, 1, 3, 1, 7, 4})) // 3,5,7 => 3
	fmt.Println(lis([]int{4, 2, 3, 1, 5}))          // 2,3,5 => 3
}

File: sample/binary-search/lis.hs
-- longest-increasing-subsequence O(n^2)
import Data.List (maximumBy)

lis :: [Int] -> [Int]
lis = g where
 c a b = compare (length a) (length b)
 -- g [] = []
 g (x:xs) = x : maximumBy c (map (h . flip drop xs) [0..(length xs)]) where
  h [] = []
  h ys = if head ys > x then g ys else []

main = do
  print $ lis [3,2,5,1,3,1,7,4]  -- [3, 5, 7]
  print $ lis [1,3,2,5,4,5,6]  -- [1,2,4,5,6]